macos/ds_store.ksy

```diff
@@ -0,0 +1,202 @@
+        repeat: expr
+        repeat-expr: num_directories
+        doc: Master blocks of the different B-trees
+  block_descriptor:
+    seq:
+      - id: address_raw
+        type: u4
+    instances:
+      offset:
+        value: (address_raw & ~0x1f) + 4
+      size:
+        value: 1 << address_raw & 0x1f
+  directory_entry:
+    seq:
+      - id: len_name
+        type: u1
+      - id: name
+        size: len_name
+        type: str
+      - id: block_id
+        type: u4
+  free_list:
+    seq:
+      - id: counter
+        type: u4
+      - id: offsets
+        type: u4
+        repeat: expr
+        repeat-expr: counter
+  master_block_ref:
+    params:
+      - id: idx
+        type: u8
+    instances:
+      master_block:
+        pos: _parent.block_addresses[_parent.directory_entries[idx].block_id].offset
+        size: _parent.block_addresses[_parent.directory_entries[idx].block_id].size
+        type: master_block
+  master_block:
+    seq:
+      - id: block_id
+        type: u4
+        doc: Block number of the B-tree's root node
+      - id: num_internal_nodes
+        type: u4
+        doc: Number of internal node levels
+      - id: num_records
+        type: u4
+        doc: Number of records in the tree
+      - id: num_nodes
+        type: u4
+        doc: Number of nodes in the tree
+      - type: u4
+        doc: Unknown field
+    instances:
+      root_block:
+        io: _root._io
+        pos: _root.buddy_allocator_body.block_addresses[block_id].offset
+        type: block
+  block:
+    seq:
+      - id: mode
+        type: u4
+        doc: If mode is 0, this is a leaf node, otherwise it is an internal node
+      - id: counter
+        type: u4
+        doc: Number of records or number of block id + record pairs
+      - id: data
+        type: block_data(mode)
+        repeat: expr
+        repeat-expr: counter
+    instances:
+      rightmost_block:
+        io: _root._io
+        pos: _root.buddy_allocator_body.block_addresses[mode].offset
+        type: block
+        if: mode > 0
+        doc: Rightmost child block pointer
+  record:
+    seq:
+      - id: filename
+        type: ustr
+      - id: structure_type
+        type: four_char_code
+        doc: Description of the entry's property
+      - id: data_type
+        size: 4
+        type: str
+        doc: Data type of the value
+      - id: value
+        type:
+          switch-on: data_type
+          cases:
+            '"long"': u4
+            '"shor"': u4
+            '"bool"': u1
+            '"blob"': record_blob
+            '"type"': four_char_code
+            '"ustr"': ustr
+            '"comp"': u8
+            '"dutc"': u8
+  block_data:
+    params:
+      - id: mode
+        type: u4
+    seq:
+      - id: block_id
+        type: u4
+        if: mode > 0
+      - id: record
+        type: record
+    instances:
+      block:
+        io: _root._io
+        pos: _root.buddy_allocator_body.block_addresses[block_id].offset
```

Comment:
1. Why is `size` unused? Should it be?
2. I guess it may make sense to make `block` a child of `block_descriptor` and make it available via `block` instance of the descriptor.

Reply:
1. `size` is not used because otherwise I get the following error message (or a similar one):

```
Parse error: undefined
Call stack: undefined KaitaiEOFError: requested 34 bytes, but only -12 bytes available
```

This is because the whole structure is a B-tree. A block can contain other blocks which in their turn can contain other blocks, etc.

2. It's possible but if you do so, you won't see the hierarchy of the B-tree, only a flat list of blocks.

Reply:
>It's possible but if you do so, you won't see the hierarchy of the B-tree, only a flat list of blocks.

Not sure in this statement. IMHO nestedness will be preserved: nearly everywhere you use blocks, you use their descriptors. So It is descriptors who define the blocks. BTW: `value` instances can be references to structs.

Reply:
>size is not used because otherwise I get the following error message (or a similar one):

>This is because the whole structure is a B-tree. A block can contain other blocks which in their turn can contain other blocks, etc.

Largely depends on the structure of the stuff. If these sizes are meaningful, they probably can be enforced by using the right `io`

Reply:
> Not sure in this statement. `value` instances can be references to structs.

Yes but you won't be able to know which upper-level block the one you are currently looking at is linked to, only the sublevel ones. However, a `block` child value can be added to `block_descriptor` in addition to the B-tree structure starting with `master_block`.

Reply:
> Largely depends on the structure of the stuff. If these sizes are meaningful, they probably can be enforced by using the right `io`.

Certainly, but I haven't found a way to make it work so far.